### PR TITLE
Add an option to specify mock GPU topology config

### PR DESCRIPTION
**Pull request by jaro-sevcik**

### CODE REVIEW

**Comment by jaro-sevcik** [comment]
This depends on https://github.com/openxla/xla/pull/15577.

**Comment by yashk2810** [comment]
I don't think we should we adding this flag. We should instead integrate mock GPUs with jax.experimental.topology and use that interface to do this kind of thing.

This is already diverging from TPUs.

**Comment by jaro-sevcik** [comment]
We need to synchronize on this, @cheshire and @nouiz were pushing for the approach sketched in this draft.

I have been also working on the topology based approach (see a hacky prototype at https://github.com/jaro-sevcik/xla/commit/2322843685a7a4f323e1580a9216aa4b13e68008 and an example usage at https://github.com/jaro-sevcik/xla/commit/3af6c1744b49cf1437e521c7a56579ebe1febadd), but there was some resistance to that, so I've put it on hold.

**Comment by yashk2810** [comment]
> but there was some resistance to that, so I've put it on hold

Why was there resistance? JAX APIs should be the same across different runtimes (this is how all other APIs work). Atleast from what I see, there seems to be no reason to differ here? Your https://github.com/jaro-sevcik/xla/commit/3af6c1744b49cf1437e521c7a56579ebe1febadd seems great!

File: tests/mock_gpu_topology_test.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    f = jax.jit(jnp.sum,
+                in_shardings=NamedSharding(mesh, P('x')),
+                out_shardings=NamedSharding(mesh, P()))
+
+    f_lowered = f.lower(jnp.arange(16))
+    hlo = f_lowered.compiler_ir()
+
+    mocked_count = NUM_SLICES * NUM_HOSTS_PER_SLICE
+    self.assertIn(f'num_partitions = {mocked_count}', str(hlo))
+    if config.use_shardy_partitioner.value:
```
Comment (yashk2810) [inline tests/mock_gpu_topology_test.py:55]: Remove this if else? You can just test for non-shardy case?

Comment (jaro-sevcik) [reply tests/mock_gpu_topology_test.py:55]: Won't we lose coverage if shardy becomes default?

Comment (yashk2810) [reply tests/mock_gpu_topology_test.py:55]: Let's worry about that when it happens. Should be a very simple change to the test to make this test work.

I am just trying to reduce the number of configs we run internally.

Comment (jaro-sevcik) [reply tests/mock_gpu_topology_test.py]: Ah, I switched that to "gpu_h100" but happy to change to any other `gpu_2gpu` if you think that's better.

**Review by yashk2810 — APPROVED** [review]

**Pull request merged**